pkg/ccl/multiregionccl/region_test.go

```diff
@@ -495,50 +468,220 @@ CREATE TABLE db.rbr(k INT PRIMARY KEY, v INT NOT NULL) LOCALITY REGIONAL BY ROW;
+				})
+			})
+		}
+	}
+}
+
+// TestAlterRegionalByRowEnclosingRegionAddDrop tests altering a
+// (which may or may not succeed) with a concurrent operation on a regional by
+// row table. The sketch of the test is as follows:
+// - Client 1 performs an ALTER TABLE ... SET LOCALITY REGIONAL BY ROW TABLE.
+// Let the user txn commit.
+// - Block in the (table) schema changer.
+// - Client 2 performs an ALTER ADD / DROP REGION. Let the operation complete
+// (succeed or fail, depending on the particular setup).
+// - Resume the ALTER operation in the schema changer.
+// Currently, this ALTER operation is bound to fail because of
+// https://github.com/cockroachdb/cockroach/issues/64011. This test ensures that
+// the rollback completes successfully. Once the issue is addressed, this test
+// should be updated to assert the correct partitioning values (in all cases).
+func TestAlterRegionalByRowEnclosingRegionAddDrop(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	skip.UnderRace(t, "times out under race")
+
+	// Decrease the adopt loop interval so that retries happen quickly.
+	defer sqltestutils.SetTestJobsAdoptInterval()()
+
+	// If the schema change succeeds, the alter may fail. The reason for the alter
+	// failing is dependent on the particular region change operation being
+	// performed. That is why we have the `expectedErrorRe` field on this struct.
+	// See https://github.com/cockroachdb/cockroach/issues/64011 for more details.
+	regionAlterCmds := []struct {
+		name               string
+		cmd                string
+		shouldSucceed      bool
+		expectedPartitions []string
+		expectedErrorRE    string
+	}{
+		{
+			name:               "drop-region-fail",
+			cmd:                `ALTER DATABASE db DROP REGION "us-east3"`,
+			shouldSucceed:      false,
+			expectedPartitions: []string{"us-east1", "us-east2", "us-east3"},
+		},
+		{
+			name:               "drop-region-succeed",
+			cmd:                `ALTER DATABASE db DROP REGION "us-east3"`,
+			shouldSucceed:      true,
+			expectedPartitions: []string{"us-east1", "us-east2"},
+			expectedErrorRE:    "in enum \"public.crdb_internal_region\"",
+		},
+		{
+			name:               "add-region-fail",
+			cmd:                `ALTER DATABASE db ADD REGION "us-east4"`,
+			shouldSucceed:      false,
+			expectedPartitions: []string{"us-east1", "us-east2", "us-east3"},
+		},
+		{
+			name:               "add-region-succeed",
+			cmd:                `ALTER DATABASE db ADD REGION "us-east4"`,
+			shouldSucceed:      true,
+			expectedPartitions: []string{"us-east1", "us-east2", "us-east3", "us-east4"},
+			expectedErrorRE:    "missing partition us-east4 on PRIMARY INDEX of table t",
+		},
+	}
+
+	testCases := []struct {
+		name  string
+		setup string
+	}{
+		{
+			name:  "alter-from-global",
+			setup: `CREATE TABLE db.t () LOCALITY GLOBAL`,
+		},
+		{
+			name:  "alter-from-explicit-regional",
+			setup: `CREATE TABLE db.t () LOCALITY REGIONAL IN "us-east2"`,
```

Comment:
i'm convinced `REGIONAL BY ROW` -> `GLOBAL` / `REGIONAL BY TABLE` is broken in that it creates extraneous partitions / zone configs, but looks like we're not there yet....